Commands and Networks

Cogno-Marco opened the pull request:
## Premise
Even though these seem like a lot of lines, you'll see that half of those are just tests, so no need to worry.

This is the first part of the stuff we've worked on, that we're introducing into the central repository.
In this pr you'll find Commands (with some refactor) and Networks fully implemented and _mostly_ tested

## Changes

### 1. Gradle files
We've done some minor edits to the gradle files, specifically, we've updated the smslibrary and we've added mocks and powermocks to use them in tests

### 2. Abstract commands
In the folder containing the interfaces for the network, we've added the commands as abstract classes, this lets us, when we extend those, have the execute command as `protected`, so it can't be called outside of that package. This way you can't call the execute() command yourself, but you can only give it to the `CommandExecutor` (which is in the same package) 

### 3. SMSInvitation
The SMSInvitation class has been implemented. This class is used (as per specifications, see #6 ) to notify an outside user about an invitation received by an user inside a network.

### 4. NetworkManager and JoinableNetworkManager
These classes have been fully implemented using commands.

## Side changes
### 1. RemovePeer to QuitNetwork
The `RemovePeer` command has been renamed into `QuitNetwork`, this is because, if we implemented a RemovePeer command, an user could choose to remove from the net people how much he wanted, making it a vulnerability for the network. To not have that possibility, we decided to implement a command which can only remove the user who sends that command, making it effectively a `QuitNetwork` command, hence the renaming

___
Let's finish this month of updates 🚀

giovannivelludo committed:
Merge pull request #1 from Cogno-IDU/master

Merge "Cogno-IDU/network-dictionary" into "EIS0/network-dictionary"

giovannivelludo committed:
fixed .gitignore

giovannivelludo committed:
Delete modules.xml

giovannivelludo committed:
Merge pull request #2 from EIS0/hotfix/gitignore

Hotfix/gitignore

giovannivelludo committed:
Merge pull request #3 from Cogno-IDU/master

fixed .gitignore (#3)

giovannivelludo committed:
updated submodule library

giovannivelludo committed:
Merge remote-tracking branch 'upstream/master'

Cogno-Marco committed:
Merge remote-tracking branch 'central/master'

Cogno-Marco committed:
Merge remote-tracking branch 'central/master'

Cogno-Marco committed:
Added missing tests folders

Cogno-Marco committed:
Defined interface for NetworkDictionary

Cogno-Marco committed:
Improved Specifications
+ Added unit tests in SMSNetDictionaryTest
- Removed example test class
~ Improved specifications for SMSNetDictionary

Cogno-Marco committed:
Implemented all NetworkDictionary functions, every test was passed

Cogno-Marco committed:
Defined interface for NetSubscribers

Cogno-Marco committed:
Unit tests
~ Improved specifications
+ Created unit tests

Cogno-Marco committed:
Implemented SMSNetSubscribers class, every unit test was passed

Cogno-Marco committed:
Improved SMSNetDictionary constructor

Cogno-Marco committed:
Merge branch 'feature/subscribers' into develop

EnricoCestaro committed:
Added Commands

EdoardoRaimondi committed:
idk

EdoardoRaimondi committed:
Command interface added

EdoardoRaimondi committed:
CommandExecutor class added

EnricoCestaro committed:
Implemented interface

EnricoCestaro committed:
Correcting a mistake (of mine)

EnricoCestaro committed:
Basic Commands structure completed

EdoardoRaimondi committed:
Commit to move to an apposite branch

EdoardoRaimondi committed:
Merge branch 'develop' of https://github.com/EIS0/network-dictionary into develop

# Conflicts:
#	networklibrary/src/main/java/com/eis/communication/network/Commands/AddPeer.java

EdoardoRaimondi committed:
AddPeer command created

EnricoCestaro committed:
2 classes added

BroadcastSender to deliver messages to all the peers inside the network
BroadcastReceiver to receive them
A MessageBuilder could be a good idea

EdoardoRaimondi committed:
Subscribers commands added, with test

EdoardoRaimondi committed:
Add and remove resources commands added (with tests)

Cogno-Marco committed:
+ Defined specifications for SMSNetworkManager and SMSJoinableNetManager
+ implemented acceptJoin method structure using inheritance

Cogno-Marco committed:
Added missing todos

Cogno-Marco committed:
~ Moved test in proper test folder
~ Fixed InvalidTelephone exceptions in a couple of tests

Cogno-Marco committed:
Moved listener interfaces to a proper folder

Cogno-Marco committed:
Fixed import errors

EdoardoRaimondi committed:
Changes:

MessageBuilder, DictionaryMessage, SubscribersMessage classes created
RequestType enum created
Moved all the listeners to a properly package

EdoardoRaimondi committed:
Merge branch 'feature/joinable-net' of https://github.com/EIS0/network-dictionary into broadcast

EdoardoRaimondi committed:
Merge branch 'broadcast' of https://github.com/EIS0/network-dictionary into broadcast

EdoardoRaimondi committed:
clean errors(It doesn't recogize me listeners interfaces)

Cogno-Marco committed:
Revert "clean errors(It doesn't recogize me listeners interfaces)"

This reverts commit 4fa2c58daff43f61e8b9622ada9be19ac79be89f.

Cogno-Marco committed:
Fixed listeners import errors

Cogno-Marco committed:
Code review edits (by Luca Crema)
+ forced java version to 1.8
~ Corrected typo in INetSubscribers, SMSNetSubscriberTest and SMSNetSubscribers
~ Commands now require the list they need instead of the whole network manager
~ Fixed tests due to Commands updates
+ Added missing specifications in SMSNetDictionary

Cogno-Marco committed:
Merge branch 'pull-requests/basic-structure' into broadcast

# Conflicts:
#	networklibrary/src/main/java/com/eis/communication/network/GetResourceListener.java
#	networklibrary/src/main/java/com/eis/communication/network/InviteListener.java
#	networklibrary/src/main/java/com/eis/communication/network/JoinInvitationListener.java
#	networklibrary/src/main/java/com/eis/communication/network/JoinableNetworkManager.java
#	networklibrary/src/main/java/com/eis/communication/network/Listeners/GetResourceListener.java
#	networklibrary/src/main/java/com/eis/communication/network/Listeners/InviteListener.java
#	networklibrary/src/main/java/com/eis/communication/network/Listeners/JoinInvitationListener.java
#	networklibrary/src/main/java/com/eis/communication/network/Listeners/RemoveResourceListener.java
#	networklibrary/src/main/java/com/eis/communication/network/Listeners/SetResourceListener.java
#	networklibrary/src/main/java/com/eis/communication/network/ListenersInterfaces/GetResourceListener.java
#	networklibrary/src/main/java/com/eis/communication/network/ListenersInterfaces/InviteListener.java
#	networklibrary/src/main/java/com/eis/communication/network/ListenersInterfaces/JoinInvitationListener.java
#	networklibrary/src/main/java/com/eis/communication/network/ListenersInterfaces/RemoveResourceListener.java
#	networklibrary/src/main/java/com/eis/communication/network/ListenersInterfaces/SetResourceListener.java
#	networklibrary/src/main/java/com/eis/communication/network/NetworkManager.java
#	networklibrary/src/main/java/com/eis/communication/network/RemoveResourceListener.java
#	networklibrary/src/main/java/com/eis/communication/network/SMSNetworkManager.java
#	networklibrary/src/main/java/com/eis/communication/network/SetResourceListener.java

Cogno-Marco committed:
+ Added enum char data
~ Revamper SubscribersMessage

Cogno-Marco committed:
Implemented naive broadcast

Cogno-Marco committed:
Basic test class
+ Created basic test class for broadcast (with no test yet)
~ Edited some variables to reach values

giovannivelludo committed:
~ package names are now entirely lowercase

giovannivelludo committed:
~ used stricter access modifiers where possible
~ improved specifications of certain methods

giovannivelludo committed:
+ added message structure

giovannivelludo committed:
~ used an HashSet instead of ArrayList in SMSNetSubscribers to improve performance of add and remove operations;
~ made SMSJoinableNetManager a singleton;
+ wrote skeleton of method onMessageReceived in BroadcastReceiver.

giovannivelludo committed:
+ added sanity checks on sender's number

giovannivelludo committed:
+ added sanity checks on request field and key, value fields

giovannivelludo committed:
+ added another sanity check on the first field

Cogno-Marco committed:
- Removed I in Interface names
~ Renamed NetSubscribers to NetSubscriberList
+ Moved SMS implementations into proper package
~ renamed packages to smallcase

Cogno-Marco committed:
Moved SMS implementations into proper folders
~ changed folder structure to have 2 packages: one generic and one specific for an sms network
~ Fixed import errors
~ Improved some documentation for Commands
~ Improved Documentation for SMSNetDictionary key validity

Cogno-Marco committed:
~ Improved documentation for the 4 basic commands
+ Added Marco Cognolato as a commands author

Cogno-Marco committed:
Improved documentation for common Command interface

Cogno-Marco committed:
Fixed import errors

Cogno-Marco committed:
Merge branch 'pull-requests/basic-structure' into broadcast

# Conflicts:
#	networklibrary/src/main/java/com/eis/communication/network/CommandExecutor.java
#	networklibrary/src/main/java/com/eis/communication/network/NetworkManager.java
#	networklibrary/src/main/java/com/eis/communication/network/commands/AddPeer.java
#	networklibrary/src/main/java/com/eis/communication/network/commands/AddResource.java
#	networklibrary/src/main/java/com/eis/communication/network/commands/Command.java
#	networklibrary/src/main/java/com/eis/communication/network/commands/RemovePeer.java
#	networklibrary/src/main/java/com/eis/communication/network/commands/RemoveResource.java
#	networklibrary/src/main/java/com/eis/smsnetwork/SMSJoinableNetManager.java
#	networklibrary/src/main/java/com/eis/smsnetwork/SMSNetDictionary.java
#	networklibrary/src/main/java/com/eis/smsnetwork/SMSNetSubscriberList.java
#	networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java
#	networklibrary/src/test/java/com/eis/smsnetwork/commands/AddPeerTest.java
#	networklibrary/src/test/java/com/eis/smsnetwork/commands/AddResourceTest.java
#	networklibrary/src/test/java/com/eis/smsnetwork/commands/RemovePeerTest.java
#	networklibrary/src/test/java/com/eis/smsnetwork/commands/RemoveResourceTest.java

Cogno-Marco committed:
Fixed import errors

Cogno-Marco committed:
removed overgeneralyzation

Cogno-Marco committed:
Removed Overgeneralyzation

Cogno-Marco committed:
Removed overgeneralyzation

Cogno-Marco committed:
Improved author tags

Co-Authored-By: Luca Crema <lc.crema@hotmail.com>

Cogno-Marco committed:
Changed function name to reflect class name update

Co-Authored-By: Luca Crema <lc.crema@hotmail.com>

Cogno-Marco committed:
Improved RequestType construction
+ Added method to get the RequestType from the the Request string name

Cogno-Marco committed:
Updated author tags

Co-Authored-By: Luca Crema <lc.crema@hotmail.com>

Cogno-Marco committed:
~ Improved author tags for Commands
~ formatted code for Commands

Cogno-Marco committed:
+ Added nonnull final in SMSNetSubscriberList
~ formatted code on some classes
~ Improved documentation comments in SMSJoinableNetManager
~ Improved documentation comments in CommandExecutor

Cogno-Marco committed:
~ Generalized commands
~ Moved commands to different package
~ Renamed NetworkDictionary to NetDictionary

Cogno-Marco committed:
Merge branch 'pull-requests/basic-structure' into develop

# Conflicts:
#	networklibrary/src/main/java/com/eis/communication/network/CommandExecutor.java

Cogno-Marco committed:
~ Fixed package creation errors
~ Fixed function renaming errors

Cogno-Marco committed:
Merge branch 'pull-requests/basic-structure' into develop

Cogno-Marco committed:
Merge branch 'pull-requests/basic-structure' into broadcast

# Conflicts:
#	networklibrary/src/main/java/com/eis/communication/network/CommandExecutor.java
#	networklibrary/src/test/java/com/eis/communication/network/commands/AddResourceTest.java
#	networklibrary/src/test/java/com/eis/communication/network/commands/RemovePeerTest.java
#	networklibrary/src/test/java/com/eis/communication/network/commands/RemoveResourceTest.java
#	networklibrary/src/test/java/com/eis/smsnetwork/commands/AddResourceTest.java
#	networklibrary/src/test/java/com/eis/smsnetwork/commands/AddResourceTest.java~pull-requests_basic-structure
#	networklibrary/src/test/java/com/eis/smsnetwork/commands/RemovePeerTest.java
#	networklibrary/src/test/java/com/eis/smsnetwork/commands/RemovePeerTest.java~pull-requests_basic-structure
#	networklibrary/src/test/java/com/eis/smsnetwork/commands/RemoveResourceTest.java
#	networklibrary/src/test/java/com/eis/smsnetwork/commands/RemoveResourceTest.java~pull-requests_basic-structure

Cogno-Marco committed:
~ Fixed import/packages errors
- Removed broken useless files

Cogno-Marco committed:
updated Request construction

Cogno-Marco committed:
+ Added MessageBuilder Tests
~ Fixed Null Pointer Exception during message construction

Cogno-Marco committed:
~ Improved Message Builder code
~ Formatted RequestType code
~ Improved author tags in SMSJoinableNetManager, RequestType
~ Renamed SubscribersMessage into SMSNetMessageBuilder
+ Implemented some unit tests for SMSNetMessageBuilder
~ Fixed bug in SMSNetMessageBuilder/MessageBuilder

Cogno-Marco committed:
Improved specifications for SMSNetMessageBuilder

Cogno-Marco committed:
+ Added two tests to SMSNetMessageBuilderTest
+ Added NonNull annotation as part of the specifications

Cogno-Marco committed:
~ Formatted class
~ Generalized sub list and dictionary
+ Added methods to provide your implementations of a dictionary and a list

Cogno-Marco committed:
Merge branch 'pull-requests/basic-structure' into develop

Cogno-Marco committed:
Merge branch 'pull-requests/basic-structure' into broadcast

Cogno-Marco committed:
+ basic net requests implementations
~ made execute static to be used without instantiating the class

EdoardoRaimondi committed:
+ Message class

Cogno-Marco committed:
Fixed dependency in BroadcastReceiver

Cogno-Marco committed:
Merge branch 'broadcast' of https://github.com/EIS0/network-dictionary into broadcast

Cogno-Marco committed:
Basic invite SMSPeer command structure and implementation

Cogno-Marco committed:
+ Created basic SMSNetInvitation class
+ Implemented auto/user invitation handling method
~ Edited BroadcastReceiver to send the invitation to check it

Cogno-Marco committed:
+ Implemented AcceptJoin command
~ Changed from AddPeer to AcceptJoin in SMSNetworkManager

Cogno-Marco committed:
Structured basic getResource method

Cogno-Marco committed:
Basic structure method
(sorry, was not added to staging area by mistake)

giovannivelludo committed:
~ moved all commands to com.eis.communication.network.commands, made them generic and abstract, and moved their concrete implementation to com.eis.smsnetwork.smsnetcommands;
~ changed specifications of BroadcastReceiver to allow sending multiple resources or peers in a single SMSMessage

Cogno-Marco committed:
~ Fixed typo in CommandExecutor
~ Formatted code in NetDictionary
~ Improved documentation in NetDictionary
~ Fixed typos in SMSNetDictionary

Cogno-Marco committed:
Fixed more typos in SMSNedDictionary

Cogno-Marco committed:
Merge remote-tracking branch 'central/master'

Cogno-Marco committed:
Merge branch 'master' into develop

# Conflicts:
#	networklibrary/src/main/java/com/eis/communication/network/CommandExecutor.java
#	networklibrary/src/main/java/com/eis/communication/network/NetDictionary.java
#	networklibrary/src/main/java/com/eis/smsnetwork/SMSNetDictionary.java

Cogno-Marco committed:
Merge branch 'pull-requests/basic-structure' into broadcast

Cogno-Marco committed:
Merge branch 'master' into broadcast

# Conflicts:
#	.idea/misc.xml
#	networklibrary/src/main/java/com/eis/communication/network/NetSubscriberList.java
#	networklibrary/src/main/java/com/eis/communication/network/commands/AddPeer.java
#	networklibrary/src/main/java/com/eis/communication/network/commands/AddResource.java
#	networklibrary/src/main/java/com/eis/communication/network/commands/RemovePeer.java
#	networklibrary/src/main/java/com/eis/communication/network/commands/RemoveResource.java
#	networklibrary/src/main/java/com/eis/smsnetwork/SMSJoinableNetManager.java
#	networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java
#	networklibrary/src/test/java/com/eis/smsnetwork/SMSNetDictionaryTest.java
#	networklibrary/src/test/java/com/eis/smsnetwork/SMSNetSubscriberListTest.java
#	networklibrary/src/test/java/com/eis/smsnetwork/SMSNetworkManagerTest.java

Cogno-Marco committed:
~ Renamed concrete command from <name> to SMS<name>
~ Fixed broken tests
~ Moved tests to right folders

Cogno-Marco committed:
- Removed wrong generalization in NetSubscriberList
~ Fixed docs text and formatted abstract commands classes

Cogno-Marco committed:
+ Defined some Fail Reasons
+ Fully Implemented getResource

Cogno-Marco committed:
+ added Mockito and PowerMockito
+ tested getResource in SMSNetworkManager
~ changed FailReason into SMSFailReason

giovannivelludo committed:
~fixed case AddPeer in BroadcastReceiver.onMessageReceived()

EnricoCestaro committed:
Broadcast process "finished"

EnricoCestaro committed:
Merge remote-tracking branch 'origin/broadcast' into broadcast

# Conflicts:
#	networklibrary/src/main/java/com/eis/smsnetwork/broadcast/BroadcastReceiver.java

EnricoCestaro committed:
Deleted message builders

giovannivelludo committed:
~fixed cases RemovePeer and AddResource in BroadcastReceiver.onMessageReceived()

giovannivelludo committed:
Merge branch 'broadcast' of https://github.com/EIS0/network-dictionary into broadcast

# Conflicts:
#	networklibrary/src/main/java/com/eis/smsnetwork/broadcast/BroadcastReceiver.java

giovannivelludo committed:
~ fixed case RemoveResource in BroadcastReceiver.onMessageReceived()

EnricoCestaro committed:
dictionary reference added

giovannivelludo committed:
~ removed useless try-catch blocks, removed an hardwiring, simplified case RemoveResource, formatting

giovannivelludo committed:
Merge remote-tracking branch 'EIS0/broadcast' into broadcast

# Conflicts:
#	networklibrary/src/main/java/com/eis/smsnetwork/broadcast/BroadcastReceiver.java

giovannivelludo committed:
~ used Set instead of ArrayList where we need to keep track of subscribers, because it avoids duplicate elements, and makes contains() and remove() operations more efficient since it's implemented as a HashSet;
- removed class SMSNetInvitation, because its only usage could be replaced with a lambda: `() -> sender` instead of a whole class;
~ replaced String concatenations in loops with StringBuilder to improve performance;
+ added a Set of invited Peers, where they are added as they're invited, and are removed when they accept the invitation;

giovannivelludo committed:
~ fixed a test

giovannivelludo committed:
~ removed invalid VCS root mapping for directory $PROJECT_DIR$/library

Cogno-Marco committed:
+ implemented 2 basic functions in SMSNetworkManager

Cogno-Marco committed:
Merge branch 'broadcast' of https://github.com/EIS0/network-dictionary into broadcast

Cogno-Marco committed:
implemented local error management

giovannivelludo committed:
~ moved documentation of BroadcastReceiver to its only method
+ additional checks on messages where only one field should exist

giovannivelludo committed:
Merge remote-tracking branch 'EIS0/broadcast' into broadcast

giovannivelludo committed:
~ changed Strings associated to RequestTypes, now they're single GSM characters to save space in messages;
~ added tests for messages containing a valid request, but unwanted additional fields.

giovannivelludo committed:
~ renamed methods with unconventional names

EnricoCestaro committed:
Added SMSInvitation class

EnricoCestaro committed:
Merge remote-tracking branch 'origin/broadcast' into broadcast

Cogno-Marco committed:
Merge branch 'broadcast' into develop

Cogno-Marco committed:
Started little mini demo to test the network

Cogno-Marco committed:
Basic demo setup
+ List view update (hopefully)
~ Changed ScrollView to ListView

Cogno-Marco committed:
+ Added permissions
+ Added Log to see if a message is sent
~ Commented out log used for debugging purposes

Cogno-Marco committed:
~ updated SMSlibrary
+ created setup function for network
+ added some logs
~ updated library update errors

Cogno-Marco committed:
~ Fixed lambda (by removing it)
~ Fixed wrong call to listener not present
+ Added logs to debug

Cogno-Marco committed:
Added debug log

EnricoCestaro committed:
compile -> implementation

EnricoCestaro committed:
Modified the AcceptInvite case

and removed a Command in the SMSAcceptInvite class

giovannivelludo committed:
+ added tests;
~ removed an hardwiring;
- removed unused imports.

giovannivelludo committed:
Merge remote-tracking branch 'EIS0/develop' into broadcast

# Conflicts:
#	networklibrary/src/main/java/com/eis/smsnetwork/broadcast/BroadcastReceiver.java

giovannivelludo committed:
~ changed how fields in a message are separated, to allow sending keys and resources containing spaces;
+ added methods to format keys and resources so that they can be properly parsed in SMSs;
+ added methods to parse keys and resources received in SMSs, and then add them to my dictionary.

giovannivelludo committed:
+ added broadcasts in Commands

giovannivelludo committed:
- removed tests on key validity, since now spaces are accepted even in keys.

giovannivelludo committed:
fixed a test

giovannivelludo committed:
fixed a couple of tests

giovannivelludo committed:
+ method setup() now sets a custom MessageParseStrategy, needed to be able to send 160 characters long SMS messages instead of 70

giovannivelludo committed:
~ removed some hardwirings

giovannivelludo committed:
~ changed order of operations when accepting an invite, to make sure we delete our old network information before we're sent the new list of subscribers and dictionary

giovannivelludo committed:
~ fixed remaining tests

giovannivelludo committed:
~ improved documentation
+ added check for empty dictionary in getAllKeyResourcePairsForSMS()

giovannivelludo committed:
~ changed FIELD_SEPARATOR to a less frequent character, to avoid having to escape it often, thus wasting space in SMSs;
+ restored validity checks on keys and resources when added to the dictionary, they're less restrictive than before, because they required keys to not contain spaces, while now they only check that the last character is not a backslash

giovannivelludo committed:
+ added last tests for BroadcastReceiver and SMSNetDictionary

giovannivelludo committed:
~ documentation fix

giovannivelludo committed:
~ renamed RemovePeer to QuitNetwork, since it's used only by the user who wants to get out of the network;
+ added method clear() in NetDictionary to allow clearing the dictionary when switching networks;
+ added an input validation check in BroadcastReceiver;
+ when we quit a network, we now clear our subscribers list and our dictionary;
~ fixed tests;
~ fixed specifications of methods that take keys and resources as arguments in SMSNetworkManager.

giovannivelludo committed:
~ renamed occurrences of RemovePeer to QuitNetwork

Cogno-Marco committed:
fix: removed unwanted final attribute

Cogno-Marco committed:
reset: resetted main activity stuff in preparation for code review

Cogno-Marco committed:
reset: removed BroadcastReceiver in preparation for the pull-request

Cogno-Marco committed:
reset: removed RequestType in preparation for code-review

Cogno-Marco committed:
reset: removed SMSFailReason in preparation for code-review

Cogno-Marco committed:
reset: removed BroadcastSender in preparation for code-review

Cogno-Marco committed:
reset: removed BroadcastSender and Receiver Tests in preparation for code-review

Cogno-Marco committed:
reset: resetted Dictionary in preparatio for pull-request

Cogno-Marco committed:
reset: removed some tests for the dictionary in preparation for pull-request

Cogno-Marco committed:
reset: resetted subsccriber list in preparation for code-review
reset: resetted dictionary in preparation for code-review

ricdezen commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSInvitation.java:
```diff
@@ -0,0 +1,24 @@
+ * Represents a received invitation to join a network.
+ */
+public class SMSInvitation implements Invitation<SMSPeer> {
+    SMSPeer inviter;
```
Is it necessary for `inviter` to be package-private?

Cogno-Marco replied:
Im not sure. Definitely not public by looking at the constructor, but I think it might just be private. I'll carefully check with our current develop and keep you updated 👍

Cogno-Marco replied:
Okay, I checked for that variable, it can be private without any problem, so I'll fix it now 👍

ricdezen commented on app/build.gradle:
```diff
@@ -5,7 +5,7 @@ android {
     defaultConfig {
         applicationId "com.eis.networkproject"
-        minSdkVersion 15
+        minSdkVersion 26
```
What forced this big change in the minimum sdk version? I believe around 20% of devices would not be able to use the apps that require this library, more people than you might expect still have Android 6 or 7. If it's for a very solid reason then I think it could be ok.

Cogno-Marco replied:
I don't specifically remember why the sdk was increased from 15 to 26 (it happened 2-3 weeks ago afterall), but I think it's because of the smslibrary. Anyway this sdk update is in the app gradle, not the library gradle, so it shouldn't affect the library at all

ricdezen replied:
Right I didn't notice it was in the demo module, my bad

ricdezen reviewed (changes requested):
A lot of classes but the level of complexity of each is fairly low, tests are not commented but they mostly consist of single lines or have explicative names, I suggest maybe a comment to the top of the class to summarize what features are being tested and a link to the tested class.
I just need a couple of clarifications which I pointed out specifically before approving.

Cogno-Marco committed:
fix: made variable private
style: formatted code

CremaLuca commented on .idea/misc.xml:
```diff
@@ -1,5 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
+  <component name="JavadocGenerationManager">
```
Did you generate a javadoc lol?

Cogno-Marco replied:
I think I tried and failed, lol

CremaLuca commented on networklibrary/build.gradle:
```diff
@@ -35,5 +35,11 @@ dependencies {
     androidTestImplementation 'androidx.test.ext:junit:1.1.1'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'
-    implementation 'com.github.Cogno-IDU:smslibrary:v-2.0'
+    implementation 'com.github.Cogno-IDU:smslibrary:v2.2'
```
We should change the naming convention to numbers only so that android studio warns you when there's a newer version. Not related to this PR, just a side note I found out.

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/commands/AcceptInvite.java:
```diff
@@ -0,0 +1,29 @@
+ * @author Marco Cognolato
+ * @author Giovanni Velludo
+ */
+public abstract class AcceptInvite<T extends Peer> extends Command {
```
Wouldn't it be clearer to have `AcceptInvite<I extends Invitation>`? This way it would be clearer that in order to accept an invitation you need to have it in the first place, not just the address of who sent it.

Cogno-Marco replied:
🤔 Yeah that makes sense 👍

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/commands/AddPeer.java:
```diff
@@ -3,36 +3,33 @@
+ * @author Giovanni Velludo
  */
-public class AddPeer implements Command {
+public abstract class AddPeer<T extends Peer> extends Command {
```
_Thank god it's more **generic**_.

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/commands/InvitePeer.java:
```diff
@@ -0,0 +1,30 @@
+ * @author Marco Cognolato
+ * @author Giovanni Velludo
+ */
+public abstract class InvitePeer<T extends Peer> extends Command {
```
Why not a `SendInvitation` command? Because invitations could contain more than just the invited peer, they could contain a message or the network code or whatever. By using the `InvitePeer` with the peer as a parameter you're adding a step to the process of creation of the invitation and you're forcing to have a different command for each type of different invitation (I know there will never be more than one type of invitation, but I like to think bigger and adopt the best practice).

Cogno-Marco replied:
well, then this would just be a renaming of a command, right? from `InvitePeer` to `SendInvitation`

CremaLuca replied:
Nope, I mean `SendInvitation<I extends Invitation>`, so that the invitation can be passed as parameter and already contains what it wants (eg. Peer, a string containing an invitation text, network code, etc.) and it doesn't need to be instantiated inside the command.

Cogno-Marco replied:
Oh, so doing it from outside, got it. Will do 👍

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/commands/QuitNetwork.java:
```diff
@@ -0,0 +1,33 @@
+import com.eis.communication.network.NetSubscriberList;
+
+/**
+ * Command to remove a peer from the subscribers
```
Not in line with the `execute` method: I've been struggling for a couple of seconds to understand what this command actually does if it removes somebody else or the current user.

Cogno-Marco replied:
Oh you're right, we changed what the command does and never updated this comment, fixing 👍

CremaLuca commented on networklibrary/src/main/java/com/eis/communication/network/commands/RemoveResource.java:
```diff
@@ -2,37 +2,33 @@
 
-
 /**
  * Command to remove a resource from the network dictionary
```
```suggestion
 * Command to remove a resource from the local network dictionary
```

CremaLuca replied:
After almost a month I forgot what a network dictionary actually is so let's make it clear.

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSJoinableNetManager.java:
```diff
@@ -38,6 +60,22 @@ public void acceptJoinInvitation(Invitation invitation) {
+     * @param invitation The invitation received
+     */
+    public void checkInvitation(Invitation<SMSPeer> invitation) {
+        if (invitationListener == null) {
```
Noice, like how you handled it.

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java:
```diff
@@ -11,18 +15,30 @@
 
     private NetSubscriberList<SMSPeer> netSubscribers = new SMSNetSubscriberList();
     private NetDictionary<String, String> netDictionary = new SMSNetDictionary();
+    private NetSubscriberList<SMSPeer> invitedPeers = new SMSNetSubscriberList();
```
Why did you use a `NetSubscriberList`? Invited are not "subscribers" so it might cause some confusion, wasn't an ArrayList good enough?

Cogno-Marco replied:
Yeah you're right, I'll change it 👍

giovannivelludo replied:
Where is the confusion? The name clearly says those are the invited peers. Using an ArrayList doesn't prevent duplicate peers being added to the list, while the Set we're using in NetSubscriberList prevents that. Using a HashSet also makes searches in the set perform better compared to an ArrayList.

EDIT: Actually, instead of using NetSubscriberList we might just use a Set. Since NetSubscriberList<SMSPeer> has exactly the same functions of a Set<SMSPeer> we might as well remove that class entirely.

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java:
```diff
@@ -38,41 +54,73 @@
-     * @param key                 The key identifier for the resource.
-     * @param value               The identified value of the resource.
+     * @param key                 The key identifier for the resource. It cannot have a backslash as
+     *                            its last character.
```
> It cannot have a backslash as its last character.

That is pretty specific lol

Cogno-Marco replied:
Oh damn it, I forgot to remove this.
You'll see in a later pull request, this is needed because we (and by we I mean @giovannivelludo ) changed what can and cannot be a key and a resource.
It's kinda complex and out of this pr, so don't look at this *yet*

giovannivelludo replied:
Before that we forbid keys and resources from containing a certain character (spaces) at all, so this is less restrictive. This was a side effect of some changes I made to enable sending multiple keys and resources in one SMS.

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java:
```diff
@@ -38,41 +54,73 @@
+            Log.e(LOG_KEY, "There's been an error: " + e);
+            setResourceListener.onResourceSetFail(key, value, SMSFailReason.MESSAGE_SEND_ERROR);
+        }
+        if (hasSucceeded) setResourceListener.onResourceSet(key, value);
```
What about a `return` inside `catch`? So that you're sure that it has succeeded and you can remove three lines for `hasSucceeded`.

Cogno-Marco replied:
🤦‍♂

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java:
```diff
@@ -38,41 +54,73 @@
+            Log.e(LOG_KEY, "There's been an error: " + e);
+            removeResourceListener.onResourceRemoveFail(key, SMSFailReason.MESSAGE_SEND_ERROR);
+        }
+        if (hasSucceeded) removeResourceListener.onResourceRemoved(key);
```
Same as before, I'm open to discussion if `hasSucceded` has a valid reason to exist over a `return`.

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java:
```diff
@@ -83,8 +131,16 @@ public void removeResource(String key, RemoveResourceListener<String, FailReason
+            Log.e(LOG_KEY, "There's been an error: " + e);
+            inviteListener.onInvitationNotSent(peer, SMSFailReason.MESSAGE_SEND_ERROR);
+        }
+        if (hasSucceeded) inviteListener.onInvitationSent(peer);
```
Have you heard anything from Joe? How is he doing?

Cogno-Marco replied:
Joe Mama?

CremaLuca replied:
![My disappointment is immeasurable and my day is ruined](https://i.kym-cdn.com/entries/icons/mobile/000/025/543/eca.jpg)

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java:
```diff
@@ -96,6 +152,7 @@ public void acceptJoinInvitation(Invitation invitation) {
         // N.B. this function provides an implementation for automatically joining a network.
         // while SMSJoinableNetManager uses this function by sending the request to the user
         // using a listener set by the user.
+        CommandExecutor.execute(new SMSAcceptInvite((SMSPeer) invitation.getInviterPeer()));
```
See? Would have been quicker an `execute(new SMSAcceptInvite(invitation))`

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java:
```diff
@@ -105,7 +162,9 @@ public void acceptJoinInvitation(Invitation invitation) {
-
+        for (SMSPeer sub : netSubscribers.getSubscribers())
+            list.addSubscriber(sub);
+        netSubscribers = list;
```
Wait what? This has an extra step of assignment, could have used `foreach(sub in list) netSub.addSub(sub)` and that's it.

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/SMSNetworkManager.java:
```diff
@@ -115,6 +174,33 @@ public void setNetSubscriberList(@NonNull NetSubscriberList<SMSPeer> list) {
+                context.getApplicationContext());
+        SMSMessageHandler.getInstance()
+                .setMessageParseStrategy(new MessageParseStrategy<String, SMSPeer, SMSMessage>() {
+                    private String hiddenCharacter = "¤";
```
Kind of hardwiring for the hidden character?

Cogno-Marco replied:
I actually have no idea why that character is checked...
![image](https://user-images.githubusercontent.com/22344699/73648654-4f640500-467e-11ea-8b35-3a5b93ecc2b1.png)

Cogno-Marco replied:
Oh no wait, that's not what this is doing.
Basically it's used instead of the \r for the smslibrary to still keep a gsm character set.
So idk, I wouldn't really consider it an hardwiring, but I totally see why you would think it is.

giovannivelludo replied:
@CremaLuca I don't get why this would be an hardwiring, it's the same as in [DefaultSMSMessageParseStrategy](https://github.com/Cogno-IDU/smslibrary/blob/master/smslibrary/src/main/java/com/eis/smslibrary/SMSMessageHandler.java).
The reason why we use this instead of the default character CR is that a CR gets automatically converted to a new line somewhere during transmission of the SMS.

CremaLuca replied:
Yeah I didn't mean hardwiring as something that could be asked as parameter, I just prefer the character would be placed inside a constant.

giovannivelludo replied:
Right, we should make it final

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/smsnetcommands/SMSAcceptInvite.java:
```diff
@@ -0,0 +1,33 @@
+
+    protected void execute() {
+        CommandExecutor.execute(new SMSQuitNetwork(SMSJoinableNetManager.getInstance().getNetSubscriberList()));
+        SMSJoinableNetManager.getInstance().getNetSubscriberList().addSubscriber(inviter);
```
Could have asked for a generic `JoinableNetManager` in the constructor because what if I'm not using the default SMSJoinableNetManager? I would have to re-write this command _and we don't want that_.

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/smsnetcommands/SMSAcceptInvite.java:
```diff
@@ -0,0 +1,33 @@
+    protected void execute() {
+        CommandExecutor.execute(new SMSQuitNetwork(SMSJoinableNetManager.getInstance().getNetSubscriberList()));
+        SMSJoinableNetManager.getInstance().getNetSubscriberList().addSubscriber(inviter);
+        SMSManager.getInstance().sendMessage(new SMSMessage(inviter, RequestType.AcceptInvitation.asString()));
```
_Side note_: This is reasonably fixed because we're talking about SMSs and we forced this to work over our smslibrary, no need to make it variable like the line over this.

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/smsnetcommands/SMSInvitePeer.java:
```diff
@@ -0,0 +1,38 @@
+        SMSMessage messageToSend = new SMSMessage(peerToInvite, message);
+        SMSManager.getInstance().sendMessage(messageToSend);
+        Log.d("SMSINVITE_COMMAND", "Invitation Sent to: " + peerToInvite);
+        SMSJoinableNetManager.getInstance().getInvitedPeers().addSubscriber(peerToInvite);
```
[Same.](https://www.facebook.com/groups/1990044391208246/)

CremaLuca commented on networklibrary/src/main/java/com/eis/smsnetwork/smsnetcommands/SMSQuitNetwork.java:
```diff
@@ -0,0 +1,39 @@
+    protected void execute() {
+        String quitNetworkMessage = RequestType.QuitNetwork.asString();
+        BroadcastSender.broadcastMessage(netSubscribers.getSubscribers(), quitNetworkMessage);
+        SMSJoinableNetManager.getInstance().getNetSubscriberList().getSubscribers().clear();
```
A new method `SMSJoinableNetManager.clear()` might be useful for this command.

Cogno-Marco replied:
How about also asking for the SMSJoinableNetManager in the constructor? lol

CremaLuca reviewed (approved):
Where's my boy `SMSFailReason`? I was excited to see what kind of witchery you added but I can't see it in the files.

CremaLuca reviewed (changes requested):
JK, not approving, I pressed the wrong button.

CremaLuca commented:
Also big up for my man @Cogno-IDU that started using conventional commits. Keep up the good work.

Cogno-IDU commented:
> Where's my boy SMSFailReason? I was excited to see what kind of witchery you added but I can't see it in the files.

Oh shiet, well, if you want I can add it here (it's only a single little class), but we'll definitely need to discuss about it. Tell me what to do.

> Also big up for my man @Cogno-IDU that started using conventional commits.

I tried 2 styles, right now I'm using the second one.

The first was
```
+ added <feature> in <class>
- removed <feature> from <class>
~ changed <feature> in <class>
```
Which is really cool, and with a diff tool you can quickly see everything since the `+` get highlited in green, the `-` in red and the `~` in yellow, but the problem I has is it wasn't enough expressive, most of the time I would use `~` and it seemed somewhat unappropriate.

Recently I started with this other stile
```
<type>[optional scope]: <description>
```
where type is `feat, docs, style, fix, refactor, test` and basically everything I want
an example of how it comes out is
```
feat: added ...
style(SMSFailReason): formatted code
docs: improved documentation
```
Which I think is really cool.
We'll see how this goes, maybe I'll add a commit template for this

CremaLuca commented:
> if you want I can add it here

Yes, please.

Cogno-IDU commented:
> Yes, please.

Alright, let me open android studio, see you in half an eternity

Cogno-Marco committed:
Revert "reset: removed SMSFailReason in preparation for code-review"

This reverts commit 928c3904e989a9ce011881d391e6e2cfdbc97d7e.

Cogno-IDU commented:
alright, here it is

Cogno-IDU commented:
What I wanted to discuss is this.
Now that you have a bigger picture you can see
1. There are many SMSFailReasons
2. I only use a couple of them in SMSJoinableNetManager

This is because we realyzed one thing
When you want to check for an error in the network, the more work you're willing to do, and the more you're able to find the error, instead the less work you do and the more the error becomes general.

For example, if I want to add a resource to the net, if I just do a local exception handling (like I do right now) I can see if I launch an error, but if I took my time, and checked with many messages that everyone in the net got the resource, then I could more closely check for the error, for example I could check if at least one got the resource, or if at least half of subscribers got it, or even everyone, the problem is it's a lot of work, and in fact it also takes time, because I have to work with messages.

Because of this we decided to (**for now**) just work on the local side of the error, but we're more than happy to receive some help to actually work on a more informative error.

Cogno-Marco committed:
feat(AcceptInvite): changed from using Peer to Invitation, to have a better generalization

Cogno-Marco committed:
docs(QuitNetwork): fixed wrong documentation comment
docs(RemoveResource): improved documentation for what a network dictionary is
fix(SMSNetworkManager): changed from SMSNetSubscriberList to ArrayList for the peers invited but not in the network yet
improvement(SMSNetworkManager): improved code to not use a boolean
improvement(SMSNetworkManager): removed double assignment in setNetSubscriberList

Cogno-Marco committed:
fix: SMSAcceptInvite needs to get the SMSJoinableNetManager from the constructor, so it can still work with a different SMSJoinableNetManager
improvement(SMSNetworkManager): removed useless acceptJoinInvitation (has been moved to SMSJoinableNetManager)

Cogno-Marco committed:
specifications: improved both in SMSAcceptInvite and SMSInvitePeer

used in SMSInvitePeer the same trick as in SMSAcceptInvite

Cogno-Marco committed:
feat(SMSJoinableNetManager): added a clear method, to clear the state of the network
feat(SMSQuitNetwork): added the same trick where you need to provide an SMSJoinableNetManager
specifications(SMSQuitNetwork): improved specifications

Cogno-Marco committed:
feat: renamed InvitePeer to SendInvitation
fix: changed input from SMSJoinableNetManager to NetworkManager
fix: changed method to add a peer due to an array list change

Cogno-Marco committed:
fix: updated variable to final, removing a potential hardwiring
style: updated variable name as a consequence

Cogno-IDU commented:
Crema told me on DM to consider his review accepted, and what DeZen asked for was edited, so I'll consider both of those accepted 🚀
